server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1675,13 +1677,16 @@ public void preRecovery(ActionListener<Void> listener) {
         indexEventListener.beforeIndexShardRecovery(this, indexSettings, listener);
     }
 
-    public void postRecovery(String reason) throws IndexShardStartedException, IndexShardRelocatedException, IndexShardClosedException {
-        synchronized (postRecoveryMutex) {
+    public void postRecovery(String reason, ActionListener<Void> listener) throws IndexShardStartedException, IndexShardRelocatedException,
+        IndexShardClosedException {
+        SubscribableListener<Void> subscribableListener = new SubscribableListener<>();
+        postRecoveryComplete.set(subscribableListener);
```

Comment:
Hmm I preferred this as an assertion so it immediately fails in tests. If it's already set then we now get an `AlreadySetException` which might be handled by retrying the recovery.